### PR TITLE
em starterkit & em*d SoC to use device tree

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Comment by galak** [comment]
@vonhust Can you take a look.  I take what I could from the Zephyr source to come up with the dts.  Not sure I quite understand exactly what the difference is between EM7D, EM9D, and EM11D.  Are they same core different SoC, different cores, not exactly sure.

**Comment by nashif** [comment]
@galak did a quick test on em_starterkit and I am not getting any console output with this change.

**Comment by galak** [comment]
@nashif can you try it without the UART patch/commit.  See if we can limit down if its the base patch or the UART specific one causing issues.

**Comment by galak** [comment]
> @nashif can you try it without the UART patch/commit. See if we can limit down if its the base patch or the UART specific one causing issues.

Can you try the first patch by itself.  There isn't anything obvious to me about what would break serial output.  So trying to see which of the commits introduces the breakage.

**Comment by ruuddw** [comment]
The 'no console output' issue doesn't seem to have to do with the UART - at least em11d configuration worked fine for me. @nashif: I assume you tested em7d? No console output for me either - maybe related to MPU memory regions and userspace support instead of a UART issue.

File: drivers/serial/Kconfig.ns16550 [hunk 1]
```diff
@@ -73,12 +73,14 @@ config UART_NS16550_PORT_0_NAME
 	  struct.
 endif
 
+if !HAS_DTS and !ARC
```
Comment (ruuddw) [inline drivers/serial/Kconfig.ns16550]: Shouldn't the 'and' be '&&"?

File: drivers/serial/Kconfig.ns16550 [hunk 2]
```diff
@@ -132,12 +134,14 @@ config UART_NS16550_PORT_1_NAME
...
```
Comment (ruuddw) [inline drivers/serial/Kconfig.ns16550]: Same here, shouldn't the 'and' be '&&"?

Comment (galak) [reply drivers/serial/Kconfig.ns16550]: Good catch, fixed.

File: drivers/serial/Kconfig.ns16550 (hunk 1 above)
Comment (galak) [reply drivers/serial/Kconfig.ns16550]: Good catch, fixed.

**Comment by galak** [comment]
> The 'no console output' issue doesn't seem to have to do with the UART - at least em11d configuration worked fine for me. @nashif: I assume you tested em7d? No console output for me either - maybe related to MPU memory regions and userspace support instead of a UART issue.

Weird. Let me know if there's any way to get access to an em7d board and I can try and debug further.

**Comment by vonhust** [comment]
Hi Kumar

I made a simple debug on this,  the bug is found. You are right, it’s the problem of MPU configurations

You need to add one line in arc_mpu_regions.c, like following

#include <generated_dts_board.h>      // new added header file
#include <soc.h>
#include <arch/arc/v2/mpu/arc_mpu.h>
#include <linker/linker-defs.h>

The original codes rely on some definitions in autoconf.h which is generated from kconfig.

Now these definitions are moved to generated_dts_board.h

Thanks

Wayne

**Comment by vonhust** [comment]
Hi Kumar, thanks for you work, here are some my suggestions
* it's better that dts.fixups (em11d, 7d, 9d) should be in board/arc/em_starterkit. 
  The device specific definitions is em_starterkit_specific, i.e., NS16550_F0008000_xxx is only meaningful for em_starter_kit.  They should not be in arch/arc/soc.  

*  avoid uncessary duplicaitons.  The fixup information can be shared among different core configuration of emsk with different board versions.

* em7d/9d/11d.dtsi is better to be  put in the folder arc/synopsys/em_startkit/.  in the future, some new boards support will be added.

File: dts/arc/em11d.dtsi [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#include <em.dtsi>
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "snps,arcem9d";
```
Comment (ruuddw) [inline dts/arc/em11d.dtsi]: This should be arcem11d, not em9d.

**Review by ruuddw — DISMISSED** [review]
Kumar, thanks for the work. There's a few changes I think still required:
- rename 'em.dtsi' into 'emsk.dtsi' (EM is the name for the ARC core, emsk is the SoC / development board with the peripherals)
- add GPIO and SPI to the dts files, next to uart/i2c.
- minor other issues from the inline review comments

**Comment by galak** [comment]
> Hi Kumar
> 
> I made a simple debug on this,  the bug is found. You are right, it’s the problem of MPU configurations
> 
> You need to add one line in arc_mpu_regions.c, like following
> 
> #include <generated_dts_board.h>      // new added header file
> #include <soc.h>
> #include <arch/arc/v2/mpu/arc_mpu.h>
> #include <linker/linker-defs.h>
> 
> The original codes rely on some definitions in autoconf.h which is generated from kconfig.
> 
> Now these definitions are moved to generated_dts_board.h
> 
> Thanks
> 
> Wayne

Thanks for finding this, will fix it up.

File: dts/arc/em11d.dtsi (hunk 3 above)
Comment (galak) [reply dts/arc/em11d.dtsi]: Will Fix.

**Comment by galak** [comment]
> Kumar, thanks for the work. There's a few changes I think still required:
> 
> rename 'em.dtsi' into 'emsk.dtsi' (EM is the name for the ARC core, emsk is the SoC / development board with the peripherals)

Will change.

> add GPIO and SPI to the dts files, next to uart/i2c.

Will leave this for followup patches/PR

> minor other issues from the inline review comments

I assume this is the em11 dts comment, any others?

**Comment by galak** [comment]
> Hi Kumar, thanks for you work, here are some my suggestions
> 
> it's better that dts.fixups (em11d, 7d, 9d) should be in board/arc/em_starterkit.
> The device specific definitions is em_starterkit_specific, i.e., NS16550_F0008000_xxx is only meaningful for em_starter_kit. They should not be in arch/arc/soc.
> 
> avoid uncessary duplicaitons. The fixup information can be shared among different core configuration of emsk with different board versions.
> 
> em7d/9d/11d.dtsi is better to be put in the folder arc/synopsys/em_startkit/. in the future, some new boards support will be added.

I'm going to hold off on the refactorings right now.  Would be helpful if there was another board/soc port that used em{7,9,11}d to see what patterns made sense.

**Review by ruuddw — APPROVED** [review]
ok, basic EMSK dts support is now and place and verified.
I'll create an issue for the remaining enhancements.

**Pull request merged**